harness/determined/common/api/authentication.py

```diff
@@ -62,139 +69,79 @@ def default_load_user_password(
-            self.token_store.drop_user(session_user)
-            token = None
-
-        # Special case: use token provided from the container environment if:
-        # - No token was obtained from the token store already,
-        # - There is a token available from the container environment, and
-        # - No user was explicitly requested, or the requested user matches the token available
-        #   in the container environment.
-        if (
-            token is None
-            and util.get_det_username_from_env() is not None
-            and util.get_det_user_token_from_env() is not None
-            and requested_user in (None, util.get_det_username_from_env())
-        ):
-            session_user = util.get_det_username_from_env()
-            assert session_user
-            token = util.get_det_user_token_from_env()
-
-        if token is not None:
-            return UsernameTokenPair(session_user, token)
-
-        # We'll need to create a new token, so we'll need a password. If there was no requested
-        # user and we ended up falling back to the default username `determined`, then we can fall
-        # back to the default login as well. Otherwise, ask the user for their password.
-        fallback_to_default = password is None and session_user == constants.DEFAULT_DETERMINED_USER
-        if fallback_to_default:
-            password = constants.DEFAULT_DETERMINED_PASSWORD
-
-        if password is None:
-            password = getpass.getpass("Password for user '{}': ".format(session_user))
-
-        try:
-            token = do_login(self.master_address, session_user, password, cert)
-        except api.errors.ForbiddenException:
-            if fallback_to_default:
-                raise api.errors.UnauthenticatedException(username=session_user)
-            raise
-
-        self.token_store.set_token(session_user, token)
-
-        return UsernameTokenPair(session_user, token)
-
-    def is_user_active(self, username: str) -> bool:
-        return self.token_store.get_active_user() == username
-
-    def get_session_user(self) -> str:
-        """
-        Returns the session user for the current session. If there is no active
-        session, then an UnauthenticatedException will be raised.
-        """
-        return self.session.username
-
-    def get_session_token(self, must: bool = True) -> str:
-        """
-        Returns the authentication token for the session user. If there is no
-        active session, then an UnauthenticatedException will be raised.
-        """
-        if self.session is None:
-            if must:
-                raise api.errors.UnauthenticatedException(username="")
-            else:
-                return ""
-        return self.session.token
-
-
-def do_login(
-    master_address: str,
-    username: str,
-    password: str,
+def login_with_cache(
+    master_address: Optional[str] = None,
+    requested_user: Optional[str] = None,
+    password: Optional[str] = None,
```

Comment:
Maybe:
```python
def login_via_password(master, user, pass, cert, save_token = False):

def login_via_token(master, user, token, cert):
```

Reply:
Actually, now that I write this I'm feeling distressed that I don't know what it means to log in. Here's an attempt:

1. If you've come via `login`, tell master you're logging in with a `bindings.post_Login` and get a token in response
2. If you've come via `login_with_cache`:
2a. See if you can find a valid token on disk. If you can, cool, you're done.
2b. See if you can find a valid (validity still TBD) token in the environment. If you can, cool, you're done.
2c. Use your password and call `bindings.post_Login` for a token.

This tells me that `bindings.post_Login` only has effects in the validity of tokens. If you have a valid token, it doesn't matter if you "log in" or not.

I think this is a definition that can make sense, but it's a weird use of the phrase "log in". I'd guess you can be logged in to a single master in this way as many concurrent users. Also that you can port a token store or enter your token in an environment variable on a machine that's never seen a determined master and yet still be "logged in".

It also makes me wish that the expiry/validity algorithm for tokens was documented better [than this](https://docs.determined.ai/latest/reference/rest-api.html#authentication).

Reply:
"Log in" in general means you provide a username and a password and get some proof of authorization.  This is true on most systems, though it can take different forms.  We issue tokens system as proof-of-login, IMAP uses persistent connections that have logged-in state.

In a browser/HTTP context, "log in" often involves you putting your proof-of-login into a cookie, which implies that you can only be logged in as one user at a time.  This is consistent with most webpage experiences that have your one logged in username in the upper right hand corner and a "logout" button that discards or invalidates your proof-of-login information (depending on server implementation).

But in many CLIs (kubectl being a common one), you can have active sessions for many different users on many different clusters.  We support that feature as well.

So, with that definition in place:
- `login_via_password` is just "login"
- `login_via_token` doesn't make sense; the token is evidence that you are already logged in (unless the token was invalidated by the server since you received it)

> Something feels wrong about the chosen functions if a user can login_with_cache passing a bad password and yet still be successful.

I feel this concern.  I struggled with this one too.  What I decided was that it's an inevitable side-effect of having a login cache.  Imagine the following code:

```python
from determined.experimental import client

client.login(mymaster, username="myuser", password=os.environ["MY_DET_PASSWORD"])
client.create_experiment(...)
```

What if the user runs this code ten times in quick succession?  Should this code do a new login every time, just because a password was provided?  If so, what the heck good is the cache in the first place?  But if it doesn't, what if the user changes their password input?  I think you can't have it both ways.  And I think that having the cache is far more valuable than verifying the password just because it was provided.

Side note, in a secure system, a login should cost the server maybe a half-second or a second of compute resources and fairly high memory usage to validate a password.  Otherwise the passwords can be reversed out of the hashes if the database is ever leaked.  Adobe made this mistake a few years ago and got skewered hard for their mistake.  So extra logins are bad for the user experience and they're _especially_ bad for the server.

Reply:
ok -- agreed that we're not going to refactor `login_with_cache`. I'd like to leave this thread open until we've finished `authentication.py`, because I'd like to give this a last pass before putting the file behind me mentally.

Reply:
All that I have left here is a request in another comment for a little more documentation. Functionality has come to seem fine to me.